src/cargo/ops/common_for_install_and_uninstall.rs

```diff
@@ -545,37 +547,115 @@ where
             Poll::Pending => source.block_until_ready()?,
         }
     };
-    match deps.iter().map(|p| p.package_id()).max() {
-        Some(pkgid) => {
-            let pkg = Box::new(source).download_now(pkgid, config)?;
-            Ok(pkg)
+    if let Some(pkgid) = deps.iter().map(|p| p.package_id()).max() {
+        let pkg = Box::new(source).download_now(pkgid, config)?;
+        Ok(pkg)
+    } else {
+        let spe = SelectPackageError::new(dep, source);
+        bail!(anyhow::Error::from(spe))
+    }
+}
+
+struct YankedInfo {
+    package_name: InternedString,
+    source_id: SourceId,
+}
+
+struct PackageNotFound {
+    package_name: InternedString,
+    source_id: SourceId,
+    // REVIEW make it a reference or is cloning ok?
+    version_req: OptVersionReq,
+}
+
+enum SelectPackageError {
+    /// Return when the Package had been yanked from the Registry
+    Yanked(YankedInfo),
+    // Return when someone accidentally passed a git(hub|lab|ea) URL as the package name
+    UrlTreatedAsPackageName(PackageNotFound),
+    // Catch-all variant for all other errors
+    CouldntFindInRegistry(PackageNotFound),
+}
+
+fn is_package_name_a_git_url(package_name: &InternedString) -> bool {
+    if let Ok(url) = Url::parse(package_name) {
+        if let Some(domain) = url.domain() {
+            // REVIEW
+            // Are there any other git services without "git"
+            // in the domain?
+            // bitbucket?
+            // Is it possible to ask the cargo/crates team for
+            // some stats where crates projects are currently hosted on
+            return domain.contains("git");
         }
-        None => {
-            let is_yanked: bool = if dep.version_req().is_exact() {
-                let version: String = dep.version_req().to_string();
-                PackageId::new(dep.package_name(), &version[1..], source.source_id())
-                    .map_or(false, |pkg_id| source.is_yanked(pkg_id).unwrap_or(false))
-            } else {
-                false
-            };
-            if is_yanked {
-                bail!(
-                    "cannot install package `{}`, it has been yanked from {}",
-                    dep.package_name(),
-                    source.source_id()
-                )
+    }
+    false
+}
```

Comment:
Not sure where we have this in the contributing guide but we're preferring these design conversations to happen in the issue before getting to the PR stage

Reply:
Thanks for reviewing so swiftly. 

Appreciate this refactor breaches into discuss-first territory and I would have started the conversation earlier, if I had such a clear plan before starting to implement this. 

My workflow was:
add a failing test case (very easy thanks to the clear ticket)
find the function `select_dep_pkg` that throws the error
try to understand the two bails at the bottom and where my addition would fit
refactor into an enum that tries to classify failure scenarios
realise that my failure scenario requires the same info as the `SelectPackageError::CouldntFindInRegistry` variant
Implement that and get the test to pass
 
After all the tests passed, I decided to get your thoughts on the implementation. If this refactor is considered too high risk for now, I should now be able to fold my error into an else branch with a bail! macro like the existing handlers. 

Let me know

Reply:
I think the rendering of the diff might have obscured what I was referring to.  I was referring to the `REVIEW` comment, not the refactors.  Just a simple question on this in the Issue before getting to the PR stage would be preferred.

Personally, I also start a conversation on these things to help draw attention to an area of conversation to make it easier on reviewers.

Reply:
Please start a conversation in the issue about detecting cases for suggesting `--git`